### PR TITLE
fix obvious errors when NetworkManager is missing

### PR DESCRIPTION
**Pull request by fogti**
this fixes all the "startup time" errors encountered when trying to use this without NetworkManager

it doesn't really work yet tho, e.g.:
```console
$ venv/bin/eduvpn-cli interactive
Network Manager not available
keyring not available due to import not available
Traceback (most recent call last):
  File "~/devel/python-eduvpn-client/venv/lib/python3.11/site-packages/eduvpn/nm.py", line 786, in <lambda>
    GLib.idle_add(lambda: action(callback=quit_loop))
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/devel/python-eduvpn-client/venv/lib/python3.11/site-packages/eduvpn/cli.py", line 485, in update_state_callback
    state = self.nm_manager.connection_state
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "~/devel/python-eduvpn-client/venv/lib/python3.11/site-packages/eduvpn/nm.py", line 181, in connection_state
    for connection in self.client.get_active_connections()
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get_active_connections'
```

### CODE REVIEW

**Comment by jwijenbergh** [comment]
Thanks for the PR. Are you trying to use the client without NetworkManager or just looking to improve the error handling when it is not available?

Currently client cannot work without it but it does make sense to add extra error checking when NetworkManager is not available.

Note that for the error you're getting right now you might need to check if NetworkManager is available by adding an if statement here:

https://github.com/eduvpn/python-eduvpn-client/blob/master/eduvpn/cli.py#L655

**Comment by fogti** [comment]
done

**Comment by fogti** [comment]
basically, it would be nice to be able to generate plaintext configs without NM. (so that's the end goal, but this fixes just the stuff that directly crashes)

File: eduvpn/nm.py [hunk 1]
```diff
@@ -19,18 +19,16 @@
 from eduvpn.utils import run_in_glib_thread
 from eduvpn.variants import ApplicationVariant
 
-gi.require_version("NM", "1.0")  # noqa: E402
+from gi.repository import GLib  # type: ignore
```
Comment (fogti) [inline eduvpn/nm.py:22]: probably something like `gi.require_version("GLib", "2.0")` or such should also be inserted here...

Comment (jwijenbergh) [reply eduvpn/nm.py:22]: Whatever works is fine by me :)

**Comment by jwijenbergh** [comment]
Also if you're right now looking to have a very basic eduVPN cli, you may also use https://github.com/eduvpn/eduvpn-common/blob/main/cmd/cli/main.go

**Comment by jwijenbergh** [comment]
I made some tiny changes to fix #434, can you rebase?

**Comment by nezos** [comment]
@jwijenbergh if I understand correctly the changes will only skip the tests for NM? But how will it be possible with eduvpn-cli to connect without NM? Could you provide a step by step procedure?

Thanks

PS: I installed 4.2 but with existing configurations it complains about network manager - not sure if the present fix was merged in another way
```
 File "/usr/lib/python3.12/site-packages/eduvpn/nm.py", line 804, in add_connection_callback
    new_con = client.add_connection_finish(result)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
gi.repository.GLib.GError: nm-client-error-quark: NetworkManager is not running (1)
```

**Comment by jwijenbergh** [comment]
> @jwijenbergh if I understand correctly the changes will only skip the tests for NM? But how will it be possible with eduvpn-cli to connect without NM? Could you provide a step by step procedure?
> 
> Thanks
> 
> PS: I installed 4.2 but with existing configurations it complains about network manager - not sure if the present fix was merged in another way
> 
> ```
>  File "/usr/lib/python3.12/site-packages/eduvpn/nm.py", line 804, in add_connection_callback
>     new_con = client.add_connection_finish(result)
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> gi.repository.GLib.GError: nm-client-error-quark: NetworkManager is not running (1)
> ```

Hi, I think the goal is that it will save the WireGuard/OpenVPN configuration somewhere so that you can manually configure the VPN using wg-quick or the OpenVPN cli. This pr would then make the tool at least run without have the NM tools installed (so that the next step is saving the VPN config somewhere)

**Comment by fogti** [comment]
hmmm it's weird...

```
Opening browser...
2024/03/20 18:27:15 - Go - DEBUG - Getting access token
2024/03/20 18:27:15 - Go - DEBUG - Access token is not expired, returning
2024/03/20 18:27:15 - Go - DEBUG - Getting access token
2024/03/20 18:27:15 - Go - DEBUG - Access token is not expired, returning
panic: runtime/cgo: misuse of an invalid Handle

goroutine 1 [running]:
runtime/cgo.Handle.Value(...)
	/usr/lib/go/src/runtime/cgo/handle.go:124
github.com/eduvpn/eduvpn-common/types/cookie.NewWithContext({0x949398, 0xc000364090})
	/home/fogti/devel/_read-only/eduvpn-common/types/cookie/cookie.go:35 +0x13a
github.com/eduvpn/eduvpn-common/client.(*Client).InvalidProfile(0xc00009eea0, {0x949398?, 0xc000364090?}, 0xc05960?)
	/home/fogti/devel/_read-only/eduvpn-common/client/client.go:73 +0x31
github.com/eduvpn/eduvpn-common/internal/server.(*Servers).ConnectWithCallbacks(0xc00009eeb0, {0x949398, 0xc000364090}, 0xc000364060, 0x0?)
	/home/fogti/devel/_read-only/eduvpn-common/internal/server/servers.go:118 +0x16a
github.com/eduvpn/eduvpn-common/client.(*Client).GetConfig(0xc00009eea0, 0xc000115650, {0x7fff9cd9a0c0?, 0x1b?}, 0x1, 0x0?, 0x0)
	/home/fogti/devel/_read-only/eduvpn-common/client/client.go:434 +0xb05
main.getConfig(0xc00009eea0?, {0x7fff9cd9a0c0?, 0xc0000142b0?}, 0xa?)
	/home/fogti/devel/_read-only/eduvpn-common/cmd/cli/main.go:153 +0x138
main.printConfig({0x7fff9cd9a0c0, 0x1b}, 0xa?)
	/home/fogti/devel/_read-only/eduvpn-common/cmd/cli/main.go:187 +0x24a
main.main()
	/home/fogti/devel/_read-only/eduvpn-common/cmd/cli/main.go:209 +0x145
```

**Comment by jwijenbergh** [comment]
> hmmm it's weird...
> 
> ```
> Opening browser...
> 2024/03/20 18:27:15 - Go - DEBUG - Getting access token
> 2024/03/20 18:27:15 - Go - DEBUG - Access token is not expired, returning
> 2024/03/20 18:27:15 - Go - DEBUG - Getting access token
> 2024/03/20 18:27:15 - Go - DEBUG - Access token is not expired, returning
> panic: runtime/cgo: misuse of an invalid Handle
> 
> goroutine 1 [running]:
> runtime/cgo.Handle.Value(...)
> 	/usr/lib/go/src/runtime/cgo/handle.go:124
> github.com/eduvpn/eduvpn-common/types/cookie.NewWithContext({0x949398, 0xc000364090})
> 	/home/fogti/devel/_read-only/eduvpn-common/types/cookie/cookie.go:35 +0x13a
> github.com/eduvpn/eduvpn-common/client.(*Client).InvalidProfile(0xc00009eea0, {0x949398?, 0xc000364090?}, 0xc05960?)
> 	/home/fogti/devel/_read-only/eduvpn-common/client/client.go:73 +0x31
> github.com/eduvpn/eduvpn-common/internal/server.(*Servers).ConnectWithCallbacks(0xc00009eeb0, {0x949398, 0xc000364090}, 0xc000364060, 0x0?)
> 	/home/fogti/devel/_read-only/eduvpn-common/internal/server/servers.go:118 +0x16a
> github.com/eduvpn/eduvpn-common/client.(*Client).GetConfig(0xc00009eea0, 0xc000115650, {0x7fff9cd9a0c0?, 0x1b?}, 0x1, 0x0?, 0x0)
> 	/home/fogti/devel/_read-only/eduvpn-common/client/client.go:434 +0xb05
> main.getConfig(0xc00009eea0?, {0x7fff9cd9a0c0?, 0xc0000142b0?}, 0xa?)
> 	/home/fogti/devel/_read-only/eduvpn-common/cmd/cli/main.go:153 +0x138
> main.printConfig({0x7fff9cd9a0c0, 0x1b}, 0xa?)
> 	/home/fogti/devel/_read-only/eduvpn-common/cmd/cli/main.go:187 +0x24a
> main.main()
> 	/home/fogti/devel/_read-only/eduvpn-common/cmd/cli/main.go:209 +0x145
> ```

Could you try the latest stable release? `git checkout 1.2.0`

**Comment by fogti** [comment]
stable release works.

**Comment by jwijenbergh** [comment]
> stable release works.

Thanks I am aware of the dev issue. Will sort that after the new release

**Comment by jwijenbergh** [comment]
Hi, do you still need this @fogti. I think I want to implement non-networkmanager support by adding a flag to dump the config file instead.

**Comment by fogti** [comment]
as long as that part works without having network manager installed, I'm fine with it.

**Comment by jwijenbergh** [comment]
> as long as that part works without having network manager installed, I'm fine with it.

Thanks, is it okay if I close this and work on this later?

**Pull request closed without merging**